0005-R-pubsub-delayed-scheduling.md

```diff
@@ -0,0 +1,156 @@
+```
+
+Upon receiving a delayed message, the Dapr runtime will examine the due date for the message and publish it to the target topic if time is due. If the time isn't due, the Dapr runtime will hold the message until the time is right to send it without ACKing back to the broker. If the Dapr instance crashes, the message will remain in the broker and be consumed by a different instance.
+
+#### Configuring delayed messaging
+
+Users can enable delayed messaging by adding the following pub/sub component metadata (other fields specific for Redis removed for brevity):
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: pubsub
+spec:
+  type: pubsub.redis
+  version: v1
+  metadata:
+  - name: "enableDelayedMessaging"
+    value: true
+```
+
+If not set, the default value is `false` and the feature is bypassed completely.
+
+
+#### Consuming delayed messages
+
+On startup, the Dapr runtime will subscribe to a topic in the following format: `<app-id>-delayed`. This topic holds all delayed messages for the given application, and has the following benefits:
+
+1. Topics messages are scoped per application and not per component
+2. Enables infrastructure transparency in that operators can examine the underlying broker, discover the delayed messages topic and optionally look into the message queue if supported
+3. Of secondary importance, this enables users to optionally consume messages directly from the configured topic for advanced scenarios
+
+##### Interaction with namespaced consumer groups
+
+If [namespaced consumer groups](https://docs.dapr.io/developing-applications/building-blocks/pubsub/howto-namespace/#with-namespace-consumer-groups) are enabled, the topic format will result in `<app-id><namespace>-delayed` to allow for a delayed message topic that is scoped both to the app id and the namespace it is running in.
+
+#### Publishing delayed messages
+
+When Dapr receives a message that has a metadata key of `scheduledSendTime`, it will publish it to the delayed message topic in the format described above, if `enableDelayedMessaging` has been enabled on the component.
+Since this feature supports all pub/sub components, Dapr will return an error if a metadata key is present and `enableDelayedMessaging` is false or not set on the component.
+
+### Message format
+
+Dapr instances that receive a delayed message need to know the topic and other request level details (content type, etc.). To faciliate this requirement and support both CloudEvents and raw payloads, Dapr will publish a well known, typed proto message to the delayed messages topic in the following structure:
+
+```proto
+message DelayedMessage {
+  string topic = 1;
+  bytes message = 2;
+  string contentType = 3;
+  map<string, string> metadata = 4;
+  bool bulkPublish = 5;
+}
+```
+
+This provides the Dapr runtime all the information it needs to create a publish request and send it to the message broker.
+
+#### Bulk publish support
```

Comment:
@mukundansundar please verify this is feasible.

Reply:
A `BulkPublishRequest` contains multiple `BulkMessageEntry` entries. 
So if my understanding of this section is correct, you want to store the byte serialized form `BulkMessageEntry` in the` message` field in `DelayedMessage`. If that is the case, why store the `BulkMessageEntry` object? Can't we not deserialize when publishing to the delayed intermediate topic and only store the data and metadata associated with it? 

The main part of the BulkPublish API is the optimization between App and Dapr and Component Service. 

Even if the component does not support native BulkPublish support, can we send all delayed messages as a BulkPulblish request making use of the default BulkPublish route?

Reply:
> Even if the component does not support native BulkPublish support, can we send all delayed messages as a BulkPulblish request making use of the default BulkPublish route?

So that's the idea, once Dapr has received a delayed message (that contains all message entries), it will publish it using the BulkPublish route.